### PR TITLE
Position Control Refactoring Part 5

### PR DESCRIPTION
**Pull request by MaEtUgR**
**Describe problem solved by this pull request**
This is part 5 of the refactoring from the monster pr #12072 and should be easy to understand and review since the diff is manageable.

**Describe your solution**
- Switched naming and definition order of the `PositionControl` library to a clear interface best shown in the unit test for it: https://github.com/PX4/Firmware/blob/6bd543156eab2ccfe639e90d9214e2e0128daa09/src/modules/mc_pos_control/PositionControl/PositionControlTest.cpp#L105-L109 with better doxygen descriptions and internal naming.
- Simplified constraints checks above global limit e.g. `!(constraints.tilt < _lim_tilt)` to `(constraints.tilt > _lim_tilt)`.
- When working on https://github.com/PX4/Firmware/pull/13224 I was told we should rather write full include paths if they are in a subfolder of the same module such that it's more clear where the libreries are located even though cmake can do it all for you. That's why I also adapted the position control local includes to e.g. `#include "PositiopnControl/PositionControl.hpp"`.
- Added a takeoff unit test for basic functionality of states and ramp. It's harder to cover all corner cases but at least we see if it directly breaks.
- Replaced all  `= Vector3f(NAN, NAN, NAN);` with `.setNaN();`, initialized all `FlightTask` members with `{}`.
- Simplified `thrustToAttitude()` implementation.
- `constraints.speed_xy` is not used in the controller for a long time so I remove the processing to avoid confusion. Once tasks don't use it internally for communicating speeds accross inheritance anymore we can get rid of the message field.
- Remove control flags logic for correct logging, adding unit tests to make sure it's still correct.
- Removing the `skip_controller` flag is not yet possible with the old `_interfaceMapping()` but this will be replaced in the next pr.

**Test data / coverage**
SITL hover goto tested and unit tests pass.

**Additional context**
Part 4 that goes before this: #13347

### CODE REVIEW

File: src/modules/mc_pos_control/PositionControl/PositionControl.cpp [hunk 1]
```diff
@@ -113,7 +113,6 @@ void PositionControl::setConstraints(const vehicle_constraints_s &constraints)
 void PositionControl::update(const float dt)
 {
 	if (_skip_controller) {
-
 		// Already received a valid thrust set-point.
```
[the remark this answers is not in the document]
Comment (MaEtUgR) [reply src/modules/mc_pos_control/PositionControl/PositionControl.cpp]: I could but this snippet will be gone in the next step so I didn't want to make an effort:
https://github.com/PX4/Firmware/pull/12072/files#diff-dcd121f41884535de662c42a30c6f928R107-R120

[the remark this answers is not in the document]
Comment (MaEtUgR) [reply src/modules/mc_pos_control/PositionControl/PositionControl.cpp]: There will be one place at the end of the sequence where thrust is limited for all cases and that one I can then break out into a function, good input 👍 :
https://github.com/PX4/Firmware/compare/position-control-refactoring5...acceleration-based-input#diff-dcd121f41884535de662c42a30c6f928R151-R158

**Comment by MaEtUgR** [comment]
I rebased without conflicts on master.

**Comment by bresch** [comment]
@PX4/testflights can you please test this PR in all flight modes on several multocopters?
Thanks in advance!

File: src/modules/mc_pos_control/PositionControl/ControlMath.hpp [hunk 2]
```diff
@@ -47,12 +47,18 @@ namespace ControlMath
...
+ * @param att_sp attitude setpoint to fill
  */
-void thrustToAttitude(const matrix::Vector3f &thr_sp, const float yaw_sp,
-		      vehicle_attitude_setpoint_s &attitude_setpoint);
+void thrustToAttitude(const matrix::Vector3f &thr_sp, const float yaw_sp, vehicle_attitude_setpoint_s &att_sp);
+/**
+ * Converts a body z vector and yaw set-point to a desired attitude.
+ * @param body_z a world frame 3D vector in direction of the desired body z axis
+ * @param yaw_sp the desired yaw setpoint
+ * @param att_sp att_sp attitude setpoint to fill
```
Comment (MaEtUgR) [inline src/modules/mc_pos_control/PositionControl/ControlMath.hpp]:
```suggestion
 * @param att_sp attitude setpoint to fill
```

Comment (MaEtUgR) [reply src/modules/mc_pos_control/PositionControl/ControlMath.hpp]: Fixed in place: https://github.com/PX4/Firmware/compare/08c397a3c528f80e4d26bcb9f6c1567fd2fdeae3..0d2e3463adf676661ebd46ca353df4d969506111

File: src/modules/mc_pos_control/PositionControl/PositionControl.hpp [hunk 3]
```diff
@@ -121,66 +120,42 @@ class PositionControl
...
-				vel_sp(i) = _vel_sp(i);
-
-			} else {
-				vel_sp(i) = NAN;
-			}
-		}
-
-		return vel_sp;
-	}
+	void resetIntegral() { _vel_int = matrix::Vector3f(); }
```
Comment (MaEtUgR) [inline src/modules/mc_pos_control/PositionControl/PositionControl.hpp]:
```suggestion
	void resetIntegral() { _vel_int.setZero(); }
```

Comment (MaEtUgR) [reply src/modules/mc_pos_control/PositionControl/PositionControl.hpp]: Fixed in place: https://github.com/PX4/Firmware/compare/0d2e3463adf676661ebd46ca353df4d969506111..312e10e38d2bf7a891ef24f9afef6c3ccbd7f81a

**Comment by Junkim3DR** [comment]
### Tested on NXP FMUK66 v3

**Modes Tested**
- Position Mode: Good.
- Altitude Mode: Good.
- Stabilized Mode: Good.

**Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode.

**Notes**
No issues noted, good flight in general.

**Logs**
https://review.px4.io/plot_app?log=920329d4-7468-435a-b258-17a4cfcb860e





**Modes Tested**
- Mission Plan Mode (Automated): Good.
- RTL (Return To Land): Good.

**Procedure**
Arm and Take off in mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint see landing behavior.

**Notes**
Good flight in general.

**Logs**
https://review.px4.io/plot_app?log=cc006579-0704-4735-a0fb-d88350890309






**Modes Tested**
- Position Mode: Good.
- Mission Plan Mode (Automated): Good.
- RTL (Return To Land): Good.

**Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoint see landing behavior.

**Notes**
Good flight in general.

**Logs**
https://review.px4.io/plot_app?log=bf4ef893-d4c3-4c22-bdd9-f63f0f3654d1

**Comment by dannyfpv** [comment]
tested on pixhawk4 v5 f450
position mode: no issues
mission mode: no issues
altitude mode: no issues
stabilize mode: no issues

log:
https://review.px4.io/plot_app?log=fef4dbeb-5b69-4b3d-9630-82258cf2f898

**Comment by jorge789** [comment]
**Tested on PixRacer V4:**
### Maximum wind 10.2 m/s
**Modes Tested**

Position Mode: Good.
Altitude Mode: Good.
Stabilized Mode: Good.
Mission Plan Mode (Automated): Good.
RTL: Good.

**- Procedure**
Arm and Take off in position mode, after flying for approximately one minute, switched to altitude then stabilized mode proceed to switch to mission plan mode then make sure that vehicle follows all waypoints as shown in QGC, once completed all waypoin activate RTL and see landing behaviour.

**Notes:**
When the vehicle was landing, it did not descend vertically, drifted away with the wind and landed the vehicle manually

**Log:**  https://review.px4.io/plot_app?log=1221ea62-3709-4c9d-bb9a-943495a3ccc3

**Log**  https://review.px4.io/plot_app?log=d6764bd5-f1e4-48cd-a8aa-79828a80ab2e

**Comment by bresch** [comment]
> When the vehicle was landing, it did not descend vertically, drifted away with the wind and landed the vehicle manually

@MaEtUgR This is because the tilt is limited to 12 degrees during the whole landing phase. Should be do like for the horizontal and vertical speed: gradually limit the tilt between alt1 and alt2 ?

**Comment by MaEtUgR** [comment]
> This is because the tilt is limited to 12 degrees during the whole landing phase. Should be do like for the horizontal and vertical speed: gradually limit the tilt between alt1 and alt2 ?

I know this issue. I mostly see when I do some crazy user test going full stick and shortly after takeoff it does a twitch because the tilt limit jumps. The problem is we can't get rid of the tilt limit, that is worse I figured out. Let's do something gradual but I'm fairly confident it's unrelated to this pr.
